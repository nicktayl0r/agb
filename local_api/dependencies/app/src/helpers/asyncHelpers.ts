// vue-jest seems to not like doing this in a .vue singe file component, so I'll do it here
export const LoadingBar = () => import(
  /* webpackChunkName: "LoadingBar" */
  "@/components/LoadingBar.vue"
);

export const App = () => import(
  /* webpackChunkName: "App" */
  "@/App.vue"
);

import LoadingBarClass from "@/components/LoadingBar.vue";

let loadBarInst: LoadingBarClass | undefined = undefined;
export function setLoadingBar(bar: any) {
  loadBarInst = bar;
}

export function getLoadingBar(): LoadingBarClass | undefined {
  return loadBarInst;
}


let mainApp: any | undefined = undefined;
export function setMainApp(app: any) {
  mainApp = app;
}

export function getMainApp(): any | undefined {
  return mainApp;
}

