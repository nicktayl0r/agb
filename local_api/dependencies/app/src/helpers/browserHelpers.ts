//import { gizmoLogSecurely, logError } from "@/helpers/debugHelpers";

export function isChrome() {
  const val = navigator.userAgent.includes("Chrome")
    && !navigator.userAgent.includes("Edge");
  //logMessage("is Chrome?", val);
  return val;
}

export function isFirefox() {
  const val = navigator.userAgent.includes("Firefox");
  //logMessage("is Firefox?", val);
  return val;
}

//Expanded this check based on changes to userAgent string in iOS13
//Based on solution: https://stackoverflow.com/a/58064481
export function isMobileSafari() {
  const val = (navigator.userAgent.includes("iPad")
    || navigator.userAgent.includes("iPhone")
    || (navigator.platform === "MacIntel" && navigator.maxTouchPoints > 1))
    //@ts-ignore
    && !window.MSStream;
  //logMessage("is mobile Safari?", val);
  return val;
}

export function isSafari() {
  const val = navigator.userAgent.includes("Safari")
    && !navigator.userAgent.includes("Chrome");
  //logMessage("is Safari?", val);
  return val;
}

export function canAccessTopWindow() {
  try {
    return (window.top.location && !window.top.location.href.includes("Exception"));
  } catch (e) { return false; }
}

export function beforeUnloadEvent() {
  if (isMobileSafari()) return "pagehide"; // mobile Safari does not call beforeunload, but calls pagehide before unload. yay
  return "beforeunload";
}

let testLocalStorageEnabled = true;
let isLocalStorageEnabled = false;
export function canAccessLocalStorage() {
  //https://github.com/Modernizr/Modernizr/blob/master/feature-detects/storage/localstorage.js
  if (testLocalStorageEnabled) {
    testLocalStorageEnabled = false;
    try {
      const mod = "modernizr";
      localStorage.setItem(mod, mod);
      localStorage.removeItem(mod);
      isLocalStorageEnabled = true;
    } catch (e) {
      isLocalStorageEnabled = false;
    }
  }

  return isLocalStorageEnabled;
}

