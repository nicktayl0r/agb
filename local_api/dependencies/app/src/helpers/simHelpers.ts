import { SimManager } from "@/managers/simManager";

let simManager: SimManager | undefined;
export function getSimManager() {
  if (!simManager) simManager = new SimManager();
  return simManager;
}
