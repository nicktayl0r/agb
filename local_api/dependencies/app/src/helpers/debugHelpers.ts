import * as Sentry from "@sentry/browser";

import { captureSentryEvent, captureSentryException } from "@/helpers/sentryHelper";

import { getEnv } from "@/helpers/envHelpers";
import { getStore } from "@/helpers/storeHelpers";
import { getPageID } from "@/helpers/widgetHelpers";
import { readModulesState } from "@/store/modules/userData";

import { readSessionId, readAttemptId } from "@/store/modules/authData";

import {
  readProjectName,
  readPlayerGitVersion,
  readProjectGitVersion
} from "@/store/modules/appData";

import { vueApp } from "..";

import { AxiosRetryInstance } from "@/helpers/requestHelpers";

import { canAccessLocalStorage } from "@/helpers/browserHelpers";

import { pageIsUnloading } from "@/store/mutationHistoryPlugin";

let debug = false;

export function setDebug(_debug: boolean) {
  debug = _debug;
}

export function isDebug() {
  return debug || window.location.href.includes("?debug=true");
}

export function showHandbookPages() {
  return window.location.href.includes("?handbookPages=true");
}

export function showDebugLabel() {
  return !window.location.href.includes("?debugLabel=false");
}

const filterValues = [
  "elClassID",
  "response",
  "responsedata",
  "comments",
  "score",
  "progress",
  "state",
  "attemptId",
  "totalscore",
  "sessionId",
  "userActivityAttemptId",
  "userActivityResponseId"
];


function combineOptionalParams(optionalParams: any[]) {
  let returnString = "";

  if (optionalParams != undefined) {
    for (const param of optionalParams) {
      if (param != undefined) returnString += ` ${stringifyParam(param)}`;
    }
  }

  return returnString;
}

function stringifyParam(param: any) {
  if (typeof param === "object" && param !== undefined) {
    return `\n${JSON.stringify(param, undefined, 4)}\n`;
  }
  return param.toString();
}

// clean up objects for the browser log, w/o Vue getters and setters
function removeFunctionsFromObjects(params: any[]) {
  const newArray: any[] = [];
  for (const index in params) {
    if (typeof params[index] === "object" && params[index] !== undefined) {
      newArray[index] = JSON.parse(JSON.stringify(params[index]));
    } else newArray[index] = params[index];
  }
  return newArray;
}

export function logMessage(message: string, ...optionalParams: any[]) {
  if (isDebug() || getEnv() === "development") {
    console.log(message, ...removeFunctionsFromObjects(optionalParams));
    // const stringifiedMessage = message + combineOptionalParams(optionalParams);
    // console.log(stringifiedMessage);
  }
}

export function logInfo(_message: string, ...optionalParams: any[]) {
  console.log(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Info,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logError(_message: string, ...optionalParams: any[]) {
  console.error(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Error,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logWarning(_message: string, ...optionalParams: any[]) {
  console.warn(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Warning,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logException(error: any, ...optionalParams: any[]) {
  console.error(error, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const stringifiedError = error;
    captureSentryException(stringifiedError);
  }
}

export function gizmoLogSecurely(message: string, ...optionalParams: any[]) {
  console.log(message);
  sendGizmoLogForm("INFO", message, optionalParams);
}

export function logWarningSecurely(message: string, ...optionalParams: any[]) {
  console.warn(message);

  if (!pageIsUnloading()) {
    sendGizmoLogForm("WARN", message, optionalParams);

    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: `[GizmoLog]${message}`,
      level: Sentry.Severity.Warning,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logExceptionSecurely(error: any, ...optionalParams: any[]) {
  console.error(error);

  if (!pageIsUnloading()) {
    const stringifiedError = error;
    sendGizmoLogForm("ERROR", stringifiedError, optionalParams);
    captureSentryException(new Error(`[GizmoLog]${stringifiedError}`));
  }
}

export function logErrorSecurely(message: string, ...optionalParams: any[]) {
  console.error(message);
  if (!pageIsUnloading()) {
    sendGizmoLogForm("ERROR", message, optionalParams);

    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: `[GizmoLog]${message}`,
      level: Sentry.Severity.Error,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

async function sendGizmoLogForm(
  messageType: string,
  message: string,
  ...optionalParams: any[]
) {
  //send form request
  // URL: https://api.explorelearning.com/api/v3/write-to-log
  // Staging  URL: https://test-api.explorelearning.com/api/v3/write-to-log

  let url = undefined;
  switch (getEnv()) {
    case "production":
      url = "https://api.explorelearning.com/api/v3/write-to-log";
      break;
    case "stage":
      url = "https://test-api.explorelearning.com/api/v3/write-to-log";
      break;
    // case "development":
    //   url= "http://dev-api.explorelearning.com/api/v3/write-to-log";
    //   break;
    // case "debug":
    //   //url= "http://dev-api.explorelearning.com/api/v3/write-to-log";
    //   url= "";
    //   break;
    default:
      url = "https://api.explorelearning.com/api/v3/write-to-log";
      break;
  }

  const bodyFormData = new FormData();

  try {
    bodyFormData.append("MessageType", messageType);
    bodyFormData.append(
      "Message",
      `${message} ${combineOptionalParams(optionalParams)}
      AttemptID: ${readAttemptId(getStore())} SessionID: ${readSessionId(getStore())}
      Project: ${readProjectName(getStore())}
      ProjectGitVersion: ${readProjectGitVersion(
        getStore()
      )} PlayerGitVersion: ${readPlayerGitVersion(getStore())}`
    );
    bodyFormData.append("Key", "SoylentGreen1974");
    bodyFormData.append("Subsystem", `STEMCase_VueApp_${getEnv()}`);
  } catch (error) {
    logError(`loadSimState error: ${error}`);
  }

  const headers = {
    config: { headers: { "Content-Type": "multipart/form-data" } }
  };

  if (url) {
    AxiosRetryInstance
      .post(url, bodyFormData)
      .catch((errors): any => {
        console.error("Cannot Use Gizmo Logs", errors);
        // commented below out because it was hogging sentry with junk errors
        // logError("Cannot Use Gizmo Logs", errors, headers);
      });
  }
}
