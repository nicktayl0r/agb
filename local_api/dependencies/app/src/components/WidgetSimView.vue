<template>
	<div class="simView" v-bind:class="{unsetZIndex:showRetry()}" :id="id" v-show="visibleStore" :scenename="scenename" :simstateid="simStateIDStore" :play-state="playStateStore" :muted="mutedStore">
		<loading-bar v-show="getShowProgressBar" ref="loadingBar" />
		<div class="simStateRetry" v-show="showRetry()">
			<div class="textBackgroundBasic dropShadow" v-bind:class="{ smallSimRetry:!showRetryText() }">
				<p v-show="showRetryText()">There was an issue loading this simulation. Please wait a moment and try again.</p>
				<span class="buttonParent buttonBasic-Grey buttonBlinkBlue marginTop2">
					<fieldset id="retrySimStateButton" class="button" @click="clickedRetry">
						<span class="bgImage"></span>
						<span class="slot">
							<p>Retry</p>
						</span>
					</fieldset>
				</span>
			</div>
		</div>
	</div>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";
import Widget from "./Widget";
import { LoadingBar } from "@/helpers/asyncHelpers";
import WidgetPlayable, { state } from "@/mixins/WidgetPlayable";

import { logError, logMessage, logWarning } from "@/helpers/debugHelpers";
import { getSimManager } from "@/helpers/simHelpers";
import { runWidgetEventConditions } from "@/helpers/widgetHelpers";
import { readAudioMuted } from "@/store/modules/userData";
import { ConditionsList } from "@/data models/conditionModels";

@Component({ mixins: [WidgetPlayable], components: { LoadingBar } })
export default class WidgetSimView extends Widget {
	@Prop()
	scenename: string;

	@Prop()
	simstateid: string;

	@Prop() sceneLoadedEffects: ConditionsList;

	@Prop({ default: false })
	muted: boolean;

	sceneLoaded = false;
	holdSimManagerInstructions: boolean = true;
	beforeCreate() {
		this.holdSimManagerInstructions = true;
	}

	mounted() {
		this.holdSimManagerInstructions = false;

		// this.$nextTick(function() {
		logMessage(
			"widgetSim mounted. sceneName: ",
			this.scenename,
			", simstateid: ",
			this.simstateid
		);
		//get the $el and then send the info to the simManager
		this.subscribeToSimManagerEvents();
		if (this.$el !== undefined) {
			getSimManager().AppendCanvasToElement(
				this.$el as HTMLElement,
				this.scenename
			);
			window.addEventListener("resize", this.handleResize, false);
			if (this.notGrapes) {
				this.sceneLoaded = getSimManager().GetSimInstIsLoaded(this.scenename);

				if (this.sceneLoaded) {
					runWidgetEventConditions(
						this,
						this.$props.sceneLoadedEffects,
						this.$props.id
					);
				}
			}
		} else logError("no $el on simView");

		this.DoMute(this.mutedStore);
		this.loadSimState(this.simStateIDStore);
	}

	handleResize() {
		getSimManager().UpdateCanvasSize(this.$el, this.scenename);
	}

	beforeDestroy() {
		window.removeEventListener("resize", this.handleResize, false);
		getSimManager()
			.terminateLoadSimState(this.scenename, this.owner)
			.catch((err) => {
				logError(err);
			});
		this.unsubscribeFromSimManagerEvents();
		getSimManager().UnHookSimUpdate(this.scenename);
	}

	updated() {
		// logMessage("------widgetSim updated---------");
		if (this.$el && (this.propStoreOverride("visible") as boolean)) {
			logMessage(
				"simViewWidget (",
				this.id,
				") updated, refreshing canvas size."
			);
			getSimManager().UpdateCanvasSize(this.$el, this.scenename);
		}
	}

	onAnimationDoneEvent(s: any, sN: any) {
		if (sN === this.scenename) {
			this.updateWidget("playState", "default"); //if set to "stop" then stopEffects is called.
			runWidgetEventConditions(this, this.$props.endEffects, this.$props.id);
		}
	}

	onAnimationLoopEvent(s: any, sN: any) {
		if (sN === this.scenename) {
			runWidgetEventConditions(this, this.$props.loopEffects, this.$props.id);
		}
	}

	onPreloadProgressEvent(sN: any, value: any) {
		if (sN === this.scenename) {
			// logMessage("simWidget has the loading value of ", value)
			try {
				// progress bar can fail to animate between two pages loading the same sim, where the first page hasn't finished loading the sim
				// catch the error so loading progress can continue
				const { loadingBar } = this.$refs;
				(loadingBar as any).progressBar.animate(value);
			} catch (err) {}
			if (value >= 1) {
				this.sceneLoaded = true;
				runWidgetEventConditions(
					this,
					this.$props.sceneLoadedEffects,
					this.$props.id
				);
			}
		}
	}

	subscribeToSimManagerEvents() {
		getSimManager().OnAnimationDone.subscribe(this.onAnimationDoneEvent);
		getSimManager().OnAnimationLoop.subscribe(this.onAnimationLoopEvent);
		getSimManager().OnPreloadProgress.subscribe(this.onPreloadProgressEvent);
	}

	unsubscribeFromSimManagerEvents() {
		getSimManager().OnAnimationDone.unsubscribe(this.onAnimationDoneEvent);
		getSimManager().OnAnimationLoop.unsubscribe(this.onAnimationLoopEvent);
		getSimManager().OnPreloadProgress.unsubscribe(this.onPreloadProgressEvent);
	}

	previousSimState = "";

	get simStateIDStore() {
		if (!this.notGrapes) return "";

		const simStateID = this.propStoreOverride("simstateid") as string;
		//logMessage("WidgetPlayable.get simStateIDStore. return: ", sState);
		if (simStateID != this.previousSimState) {
			logMessage(
				"WidgetSimView.get simStateIDStore. loadSimState: ",
				simStateID
			);
			this.previousSimState = simStateID;
			this.loadSimState(simStateID);
		}
		return simStateID;
	}

	get getShowProgressBar() {
		return this.notGrapes ? !this.sceneLoaded : false;
	}

	get owner() {
		return {
			PageID: this.pageID,
			WidgetID: this.id,
			SceneName: this.scenename,
			SimStateID: this.simStateIDStore,
		};
	}

	DoPlayState(s: state) {
		logMessage("WidgetSimView.DoPlayState() has been overridden");
		this.changeSimStatePlayState(s);
	}

	DoStopState(s: state) {
		logMessage("WidgetSimView.DoStopState() has been overridden");
		this.changeSimStatePlayState(s);
	}

	DoPauseState(s: state) {
		logMessage("WidgetSimView.DoPauseState() has been overridden");
		this.changeSimStatePlayState(s);
	}

	get mutedStore() {
		if (!this.notGrapes) return true;
		const m = readAudioMuted(this.$store);
		this.DoMute(m);
		return m;
	}

	DoMute(m: boolean) {
		if (!this.holdSimManagerInstructions) getSimManager().ChangeMute(m);
	}

	showRetry() {
		return this.showRetryElement;
	}

	showRetryElement = false;
	retryCount = 0;

	clickedRetry() {
		const simStateID = this.propStoreOverride("simstateid") as string;

		logWarning(
			"Sim state retry after: Network Error",
			this.scenename,
			simStateID
		);
		this.showRetryElement = false;

		setTimeout(() => {
			this.retryCount++;
			this.loadSimState(simStateID);
		}, 250);
	}

	showRetryText() {
		if (this.$el != undefined) {
			return this.$el.clientWidth > 300;
		}
		return true;
	}

	//SimManager Instructions
	changeSimStatePlayState(s: state) {
		if (!this.holdSimManagerInstructions)
			getSimManager().ChangeSimStatePlayState(this.scenename, s, this.owner);
	}

	loadSimState(simStateID: string) {
		if (!this.holdSimManagerInstructions)
			getSimManager()
				.loadSimState(
					this.scenename,
					simStateID,
					this.OnSimStateRequest,
					this.OnSimStateRequestCancelled,
					this.OnSimStateRequestFailed,
					this.owner
				)
				.catch((err) => {
					logError(
						`WidgetSimView loadSimState`,
						this.scenename,
						simStateID,
						err
					);
				});
	}

	OnSimStateRequest(sceneName: string, stateID: string) {
		//@ts-ignore this is from WidgetPlayable
		this.ApplyPlayState(false, true);
		this.retryCount = 0;
	}

	OnSimStateRequestCancelled(sceneName: string, stateID: string) {
		this.retryCount = 0;
	}

	OnSimStateRequestFailed(sceneName: string, stateID: string) {
		if (this.retryCount >= 3) return;
		//don't show the modal, don't execute effects, they will have happened already
		if (this.retryCount == 2) {
			this.updateWidget("playState", "default"); //if set to "stop" then stopEffects is called.
			//run both end and loop effects
			runWidgetEventConditions(this, this.$props.endEffects, this.$props.id);
			runWidgetEventConditions(this, this.$props.loopEffects, this.$props.id);
			this.showRetryElement = false;
			this.retryCount = 0;
		} else {
			this.showRetryElement = true;
		}
	}
}
</script>

<style>
.simView {
	overflow: hidden;
}

.simStateRetry {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	height: 100%;
	width: 100%;
	padding: 6vh;
	font-size: initial;
	color: initial;
	position: relative;
	background-color: rgba(0, 0, 0, 0.75);
	z-index: 110; /* 10 behind App.vue #guideButtonsContainer */
}

.simStateRetry .textBackgroundBasic {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
}

.smallSimRetry {
	background: transparent;
	border: none;
	box-shadow: none;
}

.unsetZIndex {
	z-index: unset !important;
}
</style>
