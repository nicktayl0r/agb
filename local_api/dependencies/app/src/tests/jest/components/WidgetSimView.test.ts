import { getLocalVue, getCleanStore } from "@/tests/jest/jestSetup";
const localVue = getLocalVue();
let store = getCleanStore();

import { shallowMount } from "@vue/test-utils";
//import { WidgetBaseTests } from "@/tests/jest/components/WidgetBaseTests";
import { MixinWidgetPlayableTests } from "@/tests/jest/mixins/MixinWidgetPlayableTests";

import {
  dispatchUpdateWidget
} from "@/store/modules/userData";

import WidgetSimView from "@/components/WidgetSimView.vue";
import { getSimManager } from "@/helpers/simHelpers";
import { SimManager } from "@/managers/simManager";

//Required to mock this module, because LoadingBar.vue creates a line object
//   Progressbar.Line  doesn't have all of its required properties.
//   So WidgetSimView fails to mount
jest.mock("progressbar.js");

describe("WidgetSimView", () => {
  beforeEach(() => {
    store = getCleanStore();
  });

  //WidgetBaseTests(WidgetSimView);
  MixinWidgetPlayableTests(WidgetSimView);

  it("should return the correct value outside of grapes", () => {
    const wrapper = shallowMount(WidgetSimView, {
      store,
      propsData: {
        id: "testWidget",
        scenename: "testScene",
        simstateid: "testSimStateID"
      }
    });
    // wrapper.setData({ store: undefined });
    // should return !sceneLoaded if grapes isn't detected
    wrapper.setData({ sceneLoaded: true });
    expect((wrapper.vm as any).getShowProgressBar).toBe(false);

    wrapper.setData({ sceneLoaded: false });
    expect((wrapper.vm as any).getShowProgressBar).toBe(true);
  });

  it("should return false if within Grapes", () => {
    const wrapper = shallowMount(WidgetSimView, {
      propsData: {
        id: "testWidget",
        scenename: "testScene",
        simstateid: "testSimStateID"
      }
    });
    // should be false if Grapes is detected

    wrapper.setData({ sceneLoaded: true });
    wrapper.setData({ $store: undefined });

    expect((wrapper.vm as any).getShowProgressBar).toBe(false);
    wrapper.setData({ sceneLoaded: false });
    expect((wrapper.vm as any).getShowProgressBar).toBe(false);
  });

  it("matches html snapshot", () => {
    const wrapper = shallowMount(WidgetSimView, {
      propsData: {
        id: "testWidget"
      },
      localVue,
      store
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("Owner", () => {
    const owner = {
      PageID: "testPage",
      WidgetID: "testWidget",
      SceneName: "testScene",
      SimStateID: "testSimStateID"
    };

    const wrapper = shallowMount(WidgetSimView, {
      propsData: {
        id: owner.WidgetID,
        scenename: owner.SceneName,
        simstateid: owner.SimStateID
      },
      localVue,
      store
    });

    const { vm } = wrapper;

    //Use toStrictEqual to do a deep equality comparison
    expect((vm as any).owner).toStrictEqual(owner);

    owner.SimStateID = "different_testSimStateID";
    dispatchUpdateWidget(store, {
      pageID: "testPage",
      widgetID: "testWidget",
      key: "simstateid",
      value: "different_testSimStateID"
    });

    //wrapper.vm.$forceUpdate();
    expect((vm as any).owner).toStrictEqual(owner);
  });


  it("unmount triggers terminate", () => {
    const owner = {
      PageID: "testPage",
      WidgetID: "testWidget",
      SceneName: "testScene",
      SimStateID: "testSimStateID"
    };

    const wrapper = shallowMount(WidgetSimView, {
      propsData: {
        id: owner.WidgetID,
        scenename: owner.SceneName,
        simstateid: owner.SimStateID
      },
      localVue,
      store
    });

    const { vm } = wrapper;
    const simManager: SimManager = getSimManager();
    simManager.terminateLoadSimState = jest.fn((sceneName: string, owner: any) => { return Promise.resolve(); });
    wrapper.destroy();

    expect(simManager.terminateLoadSimState).toBeCalledWith("testScene", owner);
  });
});
