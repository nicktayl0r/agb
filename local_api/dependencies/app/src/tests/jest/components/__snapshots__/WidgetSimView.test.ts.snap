// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetSimView matches html snapshot 1`] = `
<div
  class="simView"
  id="testWidget"
  play-state="default"
>
  <div
    class="loadingBarParent"
  >
    <div
      class="loadingBar"
    >
      <div
        class="loadingBarSpinner"
      />
    </div>
  </div>
   
  <div
    class="simStateRetry"
    style="display: none;"
  >
    <div
      class="textBackgroundBasic dropShadow"
    >
      <p>
        There was an issue loading this simulation. Please wait a moment and try again.
      </p>
       
      <span
        class="buttonParent buttonBasic-Grey buttonBlinkBlue marginTop2"
      >
        <fieldset
          class="button"
          id="retrySimStateButton"
        >
          <span
            class="bgImage"
          />
           
          <span
            class="slot"
          >
            <p>
              Retry
            </p>
          </span>
        </fieldset>
      </span>
    </div>
  </div>
</div>
`;
